### PR TITLE
Issue #3555 - Do not reveal base resource in error response

### PR DESCRIPTION
**Pull request by joakime**
Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 1]
```diff
@@ -141,17 +140,15 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
                 }
                 else
                 {
                     writer.write("<li>");
                     writer.write(context.getContextPath());
                     if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
                         writer.write("&nbsp;@&nbsp;"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write("&nbsp;--->&nbsp;");
+                    writer.write("&nbsp;");
                     writer.write(context.toString());
```
Comment (joakime) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: missed this one. ugh.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 2]
```diff
@@ -141,18 +140,15 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
                     writer.write(context.getContextPath());
                     if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
                         writer.write("&nbsp;@&nbsp;"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write("&nbsp;--->&nbsp;");
-                    writer.write(context.toString());
-                    writer.write("</a></li>\n");
+                    writer.write("&nbsp;</a></li>\n");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Can you please add the webapp Name and status?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 3]
```diff
@@ -141,18 +140,15 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
                 }
                 else
                 {
                     writer.write("<li>");
                     writer.write(context.getContextPath());
                     if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
                         writer.write("&nbsp;@&nbsp;"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write("&nbsp;--->&nbsp;");
-                    writer.write(context.toString());
+                    writer.write("&nbsp;");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Ditto.

**Comment by joakime** [comment]
Removed VirtualHosts.

Current version of output looks like this ...

image

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 4]
```diff
@@ -116,62 +119,87 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+             OutputStreamWriter writer = new OutputStreamWriter(outputStream, UTF_8))
         {
-            writer.write("<HTML>\n<HEAD>\n<TITLE>Error 404 - Not Found");
-            writer.write("</TITLE>\n<BODY>\n<H2>Error 404 - Not Found.</H2>\n");
-            writer.write("No context on this server matched or handled this request.<BR>");
-            writer.write("Contexts known to this server are: <ul>");
+            writer.append("<!DOCTYPE html>\n");
+            writer.append("<html lang=\"en\">\n<head>\n");
+            writer.append("<title>Error 404 - Not Found</title>");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Missing `\n` after `</title>`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 5]
```diff
@@ -116,62 +119,87 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
 
             Server server = getServer();
             Handler[] handlers = server==null?null:server.getChildHandlersByClass(ContextHandler.class);
 
+            writer.append("<table class=\"contexts\"><thead><tr>");
+            writer.append("<th>Context Path</th>");
+            writer.append("<th>Display Name</th>");
+            writer.append("<th>State</th>");
+            writer.append("<th>LifeCycle</th>");
+            writer.append("</tr></thead><tbody>");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Missing `\n` after `<tbody>`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 6]
```diff
@@ -116,62 +119,87 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
+                // State
+                writer.append(context.getState());
+                writer.append("</td></tr>\n");
+            }
 
-            writer.write("\n</BODY>\n</HTML>\n");
+            writer.append("</tbody></table><hr/>\n");
+            writer.append("<a href=\"http://eclipse.org/jetty\"><img alt=\"icon\" src=\"/favicon.ico\"/></a>&nbsp;");
+            writer.append("<a href=\"http://eclipse.org/jetty\">Powered by Eclipse Jetty:// Server</a><hr/>\n");
+            writer.append("\n</body>\n</html>\n");
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Unnecessary `\n` before `</body>`.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 7]
```diff
@@ -116,62 +119,87 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
         response.setStatus(HttpServletResponse.SC_NOT_FOUND);
         response.setContentType(MimeTypes.Type.TEXT_HTML.toString());
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Given that you're using UTF-8, here you should use the `TEXT_HTML_UTF_8` constant instead.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 4 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: done

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 6 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: done

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 5 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: done

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 7 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: done

**Comment by joakime** [comment]
@sbordet done with changes from review.
Just a thought, should header "State" be actually "Status"? (it can only show "Available" or "Not Available" as values in tbody)

**Comment by joakime** [comment]
I should probably html encode the Display Name, ensure that things like `<` become `&lt;`, etc ...

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 8]
```diff
@@ -114,64 +117,89 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
             for (int i=0;handlers!=null && i<handlers.length;i++)
             {
+                writer.append("<tr><td>");
+                // Context Path
                 ContextHandler context = (ContextHandler)handlers[i];
+
+                StringBuilder href = new StringBuilder();
+
+                String contextPath = context.getContextPath();
+                href.append(contextPath);
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Should the `contextPath` be escaped?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 9]
```diff
@@ -114,64 +117,89 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
-                    writer.write(context.getContextPath());
-                    if (context.getVirtualHosts()!=null && context.getVirtualHosts().length>0)
-                        writer.write("&nbsp;@&nbsp;"+context.getVirtualHosts()[0]+":"+request.getLocalPort());
-                    writer.write("&nbsp;--->&nbsp;");
-                    writer.write(context.toString());
-                    writer.write("</a></li>\n");
+                    writer.append("<a href=\"").append(href).append("\">");
                 }
-                else
+                writer.append(context.getContextPath());
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Use the `contextPath` variable defined above.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 10]
```diff
@@ -114,64 +117,89 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
...
+                    writer.append("</a>");
                 }
-            }
+                writer.append("</td><td>");
+                // Display Name
 
-            writer.write("</ul><hr>");
+                if (StringUtil.isNotBlank(context.getDisplayName()))
+                {
+                    writer.append(context.getDisplayName());
```
Comment (sbordet) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Should the `displayName` be escaped?

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 8 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Escaping here would be interesting ...
For it to work, the user would have to use the URL escaped version, aka `ContextPath.setContextPath("/aaa%2Fbbb");`
So we wouldn't need to escape it here.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 9 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: But when writing the Textual form (not in a URL) then we should probably escape only the `%` HTML entity to `&#37;`, all other characters must be safe for that contextPath to even work properly.
In fact, we should probably do a proper URL safety check on `ContextHandler.setContextPath(String)` to ensure it doesn't use unsafe characters.

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java (hunk 10 above)
Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Yes!  definitely.

**Comment by joakime** [comment]
Pushed update based on review.

Current screenshot ...

default-handler-error3

File: jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java [hunk 11]
```diff
@@ -147,13 +148,11 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
                 // Context Path
                 ContextHandler context = (ContextHandler)handlers[i];
 
-                StringBuilder href = new StringBuilder();
-
                 String contextPath = context.getContextPath();
-                href.append(contextPath);
+                String href = UrlEncoded.encodeString(contextPath.substring(1), UTF_8);
```
Comment (gregw) [inline jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: This is the wrong encode to use.  `UrlEncoded` is for parameters not paths

Comment (joakime) [reply jetty-server/src/main/java/org/eclipse/jetty/server/handler/DefaultHandler.java]: Yep, I got that earlier.
The filed issue #3565 and associated PR #3567 are the correct solution, once resolved, that specific solution will be used.

**Comment by joakime** [comment]
Dropping, work completed in `jetty-9.2.x`

**Pull request closed without merging**